src/engine/QueryPlanner.cpp

```diff
@@ -175,58 +226,57 @@ std::vector<QueryPlanner::SubtreePlan> QueryPlanner::optimize(
         }
       }
 
-      // if our input is not optional, this means we still can arbitrarily
-      // optimize among our candidates and just append our new candidates.
-      if (!v[0]._isOptional) {
+      // all variables we have seen so far are considered "bound" from now on
+      // this also includes optionals. in the case of bind operations
+      // we do not have any variables yet, so this does no harm
```

Comment:
// All variables seen so far are considered bound and cannot appear as the RHS of a BIND operation. This is also true for variables from OPTIONALs (this was a bug in the previous version of the code).